### PR TITLE
ARSSTB-206: Save draft to submit answer

### PR DESCRIPTION
**Pull request by C-Thomp**
### Type of change
- [x] Non-breaking change

### Description
Jira link: https://jira.tools.tax.service.gov.uk/browse/ARSSTB-206

-save draft should save(submit) the current answered question before redirect
-currently covers the TextArea pages and string pages (where a user would be irked to have an answer not saved)

### To do
- [x] textarea
- [x] string pages

### CODE REVIEW

File: app/controllers/DescribeTheLegalChallengesController.scala [hunk 1]
```diff
@@ -68,9 +68,11 @@ class DescribeTheLegalChallengesController @Inject() (
                 updatedAnswers <-
                   Future.fromTry(request.userAnswers.set(DescribeTheLegalChallengesPage, value))
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(DescribeTheLegalChallengesPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/DescribeTheLegalChallengesController.scala:73]: should just be false here

File: app/controllers/DescribeTheSimilarGoodsController.scala [hunk 2]
```diff
@@ -68,9 +68,11 @@ class DescribeTheSimilarGoodsController @Inject() (
                 updatedAnswers <-
                   Future.fromTry(request.userAnswers.set(DescribeTheSimilarGoodsPage, value))
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(DescribeTheSimilarGoodsPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/DescribeTheSimilarGoodsController.scala:73]: false

File: app/controllers/ExplainHowPartiesAreRelatedController.scala [hunk 3]
```diff
@@ -67,9 +67,13 @@ class ExplainHowPartiesAreRelatedController @Inject() (
               for {
                 updatedAnswers <- ExplainHowPartiesAreRelatedPage.set(value)
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(ExplainHowPartiesAreRelatedPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainHowPartiesAreRelatedController.scala:72]: false

File: app/controllers/ExplainHowYouWillUseMethodSixController.scala [hunk 4]
```diff
@@ -68,9 +68,14 @@ class ExplainHowYouWillUseMethodSixController @Inject() (
                 updatedAnswers <-
                   Future.fromTry(request.userAnswers.set(ExplainHowYouWillUseMethodSixPage, value))
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(ExplainHowYouWillUseMethodSixPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainHowYouWillUseMethodSixController.scala:74]: false

File: app/controllers/ExplainReasonComputedValueController.scala [hunk 5]
```diff
@@ -68,9 +68,11 @@ class ExplainReasonComputedValueController @Inject() (
                 updatedAnswers <-
                   Future.fromTry(request.userAnswers.set(ExplainReasonComputedValuePage, value))
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(ExplainReasonComputedValuePage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainReasonComputedValueController.scala:73]: false

File: app/controllers/ExplainWhyYouChoseMethodFourController.scala [hunk 6]
```diff
@@ -68,9 +68,13 @@ class ExplainWhyYouChoseMethodFourController @Inject() (
                 updatedAnswers <-
                   Future.fromTry(request.userAnswers.set(ExplainWhyYouChoseMethodFourPage, value))
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(ExplainWhyYouChoseMethodFourPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainWhyYouChoseMethodFourController.scala:73]: false

File: app/controllers/ExplainWhyYouHaveNotSelectedMethodOneToFiveController.scala [hunk 7]
```diff
@@ -70,10 +70,17 @@ class ExplainWhyYouHaveNotSelectedMethodOneToFiveController @Inject() (
                     request.userAnswers.set(ExplainWhyYouHaveNotSelectedMethodOneToFivePage, value)
                   )
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator
-                  .nextPage(ExplainWhyYouHaveNotSelectedMethodOneToFivePage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainWhyYouHaveNotSelectedMethodOneToFiveController.scala:75]: false

File: app/controllers/ExplainWhyYouHaveNotSelectedMethodOneToThreeController.scala [hunk 8]
```diff
@@ -67,10 +67,17 @@ class ExplainWhyYouHaveNotSelectedMethodOneToThreeController @Inject() (
               for {
                 updatedAnswers <- ExplainWhyYouHaveNotSelectedMethodOneToThreePage.set(value)
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator
-                  .nextPage(ExplainWhyYouHaveNotSelectedMethodOneToThreePage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/ExplainWhyYouHaveNotSelectedMethodOneToThreeController.scala:72]: false

File: app/controllers/WhyComputedValueController.scala [hunk 9]
```diff
@@ -67,9 +67,11 @@ class WhyComputedValueController @Inject() (
               for {
                 updatedAnswers <- WhyComputedValuePage.set(value)
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(WhyComputedValuePage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/WhyComputedValueController.scala:72]: false

File: app/controllers/WhyIdenticalGoodsController.scala [hunk 10]
```diff
@@ -67,9 +67,11 @@ class WhyIdenticalGoodsController @Inject() (
               for {
                 updatedAnswers <- WhyIdenticalGoodsPage.set(value)
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(WhyIdenticalGoodsPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/WhyIdenticalGoodsController.scala:72]: false

File: app/controllers/WhyTransactionValueOfSimilarGoodsController.scala [hunk 11]
```diff
@@ -67,9 +67,13 @@ class WhyTransactionValueOfSimilarGoodsController @Inject() (
               for {
                 updatedAnswers <- WhyTransactionValueOfSimilarGoodsPage.set(value)
                 _              <- userAnswersService.set(updatedAnswers)
-              } yield Redirect(
-                navigator.nextPage(WhyTransactionValueOfSimilarGoodsPage, mode, updatedAnswers)
-              )
+              } yield saveDraft match {
+                case true => Redirect(routes.DraftHasBeenSavedController.onPageLoad(draftId))
+                case _    =>
```
Comment (paulhargrave) [inline app/controllers/WhyTransactionValueOfSimilarGoodsController.scala:72]: false

File: app/navigation/Navigator.scala [hunk 12]
```diff
@@ -104,8 +105,8 @@ class Navigator @Inject() () {
 
   private def whoAreYouRouting(userAnswers: UserAnswers): Call =
     userAnswers.get(WhoAreYouAgentPage) match {
-      case _ => RequiredInformationController.onPageLoad(userAnswers.draftId)
-      case _ => UnauthorisedController.onPageLoad
+      case Some(_) => RequiredInformationController.onPageLoad(userAnswers.draftId)
+      case None    => UnauthorisedController.onPageLoad
```
Comment (paulhargrave) [inline app/navigation/Navigator.scala:109]: Nice removal of unreachable code

File: app/views/components/SaveButtonsV2.scala.html [hunk 13]
```diff
@@ -0,0 +1,32 @@
+@*
```
Comment (paulhargrave) [inline app/views/components/SaveButtonsV2.scala.html:1]: Can be rename this to something more explicit?

maybe SaveButtonsWithToDraft?

Comment (C-Thomp) [reply app/views/components/SaveButtonsV2.scala.html:1]: I like that suggestion 👍

**Review by paulhargrave — APPROVED** [review]
Nice emoji

**Pull request merged**